### PR TITLE
🐛 fix race condition causing missing workstatuses

### PR DESCRIPTION
**Pull request by pdettori**
## Summary

Fixes race condition issues observed when creating manifestworks at the possible fastest rates, and issues related to consecutive deletion and re-creation.

More specifically:

- remove safemap used to correlate objects UIDs with applied manifest work, and just use the owner reference in each managed object to retrieve the related applied manifest work and manifest work.

- change naming scheme for work statuses, generating unique names based on manifest work name prepended by the applied manifest work UID. This avoids race conditions observed when using the same name between delete and re-create caused by the GC in the ITS.

## Related issue(s)

Fixes # https://github.com/kubestellar/kubestellar/issues/1820

### CODE REVIEW

File: pkg/agent/reconcilers.go [hunk 1]
```diff
@@ -124,65 +123,51 @@ func (a *Agent) handleAppliedManifestWork(obj runtime.Object, isBeingDeleted boo
...
 	ctx, cancel := context.WithTimeout(context.Background(), 2*time.Minute)
 	defer cancel()
 
-	// delete WorkStatus if exists, when the workload object is deleted
-	if isBeingDeleted {
-		err := a.hubClient.Get(ctx, client.ObjectKeyFromObject(workStatus), workStatus, &client.GetOptions{})
-		if err != nil && !apierrors.IsNotFound(err) {
-			return err
-		}
-		err = a.hubClient.Delete(ctx, workStatus, &client.DeleteOptions{})
```
Comment (andreyod) [inline pkg/agent/reconcilers.go:155]: If we remove this, how do we delete a `WorkStatus` when one manifest was removed from the list of manifests in a `ManifestWork`? With multi-wrapped objects we should have the ability to explicitly delete a `WorkStatus`, the GC will remove it only when the whole `ManifestWork` will be removed(maybe never).

File: pkg/agent/reconcilers.go [hunk 2]
```diff
@@ -124,65 +123,51 @@ func (a *Agent) handleAppliedManifestWork(obj runtime.Object, isBeingDeleted boo
...
 			// only update status for KS-managed (by bindingpolicies here) objects
 			// TODO remove the check for ManagedByKSLabelKeyPrefix after we switch to new transport
-			if !util.HasPrefixInMap(manifestWork.Labels, ManagedByKSLabelKeyPrefix) && !util.HasPrefixInMap(manifestWork.Labels, TransportLabelPrefix) {
-				a.logger.Info("object not managed by a KS bindingpolicy, status not updated", "object", name, "namespace", namespace)
+			if !(util.HasPrefixInMap(manifestWork.Labels, ManagedByKSLabelKeyPrefix) || util.HasPrefixInMap(manifestWork.Labels, TransportLabelPrefix)) {
+				a.logger.Info("object not managed by a KS bindingpolicy, status not updated", "object", aWork.Spec.ManifestWorkName, "namespace", namespace)
 				return nil
 			}
 
+			// this was causing race conditions
```
Comment (andreyod) [inline pkg/agent/reconcilers.go:170]: Maybe this comment should be removed if this is no longer causing race conditions.

**Comment by andreyod** [comment]
Regarding the name schema change in this PR: With current KS behavior, which uses multi object wrapping, the `WorkStatus` name between delete and re-create of one object in the manifests list will not change, because the `AppliedManifestWork` is modified and keep the same UID.

File: pkg/agent/reconcilers.go (hunk 1 above)
Comment (pdettori) [reply pkg/agent/reconcilers.go:155]: good catch @andreyod  - will bring back the delete code.

File: test/e2e/workstatus-creation.sh [hunk 3]
```diff
@@ -0,0 +1,89 @@
+#!/usr/bin/env bash
```
Comment (waltforme) [inline test/e2e/workstatus-creation.sh:1]: Maybe we can remove this file since it is covered by workstatus-crud.sh?

**Comment by waltforme** [comment]
I tried the driver with current `main` and observed missing workstatuses. Then I tried the driver with this branch and the workstatuses appear as expected. So look good to me.

There is one small thing: we may want to document the stress test and the driver in `test/e2e/README.md`.

File: test/e2e/workstatus-creation.sh (hunk 3 above)
Comment (pdettori) [reply test/e2e/workstatus-creation.sh:1]: done.

File: pkg/agent/reconcilers.go (hunk 2 above)
Comment (pdettori) [reply pkg/agent/reconcilers.go:170]: done

File: pkg/agent/reconcilers.go (hunk 1 above)
Comment (pdettori) [reply pkg/agent/reconcilers.go]: I have also added a test to verify that deletion of an object within a manifest causes the deletion of the corresponding work status

**Review by waltforme — APPROVED** [review]

**Pull request merged**